### PR TITLE
Add BATLedgerInitializer for managing subsystem initialization

### PR DESCRIPTION
**Pull request by zenparsing**
Resolves https://github.com/brave/brave-browser/issues/22586

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: vendor/bat-native-ledger/src/bat/ledger/internal/core/bat_ledger_initializer.cc [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+   public:
+    void Start() {
+      context().GetLedgerImpl()->database()->Initialize(
+          false, ContinueWithLambda(this, &Job::OnDatabaseInitialized));
+    }
+
+   private:
+    void OnDatabaseInitialized(mojom::Result result) {
+      if (result != mojom::Result::LEDGER_OK) {
+        context().LogError(FROM_HERE) << "Failed to initialize database";
```
Comment (emerick) [inline vendor/bat-native-ledger/src/bat/ledger/internal/core/bat_ledger_initializer.cc:34]: Is it worth logging `result` here?

File: vendor/bat-native-ledger/src/bat/ledger/internal/core/bat_ledger_initializer.cc [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+BATLedgerInitializer::~BATLedgerInitializer() = default;
+
+Future<bool> BATLedgerInitializer::Initialize() {
+  if (!initialize_future_) {
+    initialize_future_ = SharedFuture(context().StartJob<InitializeAllJob>());
+  }
+
+  return initialize_future_->Then(
+      base::BindOnce([](const bool& success) { return success; }));
```
Comment (emerick) [inline vendor/bat-native-ledger/src/bat/ledger/internal/core/bat_ledger_initializer.cc:112]: Should this just be a `bool` or is there some reason to pass it like this?

File: vendor/bat-native-ledger/src/bat/ledger/internal/core/bat_ledger_initializer.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
```
Comment (emerick) [inline vendor/bat-native-ledger/src/bat/ledger/internal/core/bat_ledger_initializer.h:1]: Nit: 2022?

File: vendor/bat-native-ledger/src/bat/ledger/internal/core/bat_ledger_initializer.cc (hunk 2 above)
Comment (zenparsing) [reply vendor/bat-native-ledger/src/bat/ledger/internal/core/bat_ledger_initializer.cc:112]: `initialize_future_` is a `SharedFuture`, which always calls its callback with a const ref (similar to how `std::shared_future::get` returns a const ref). It looks a little strange in the `bool` case, but I think I'd rather it look strange than obscure the "sharedness".


On `SharedFuture`, originally I had a template class called `FutureCache` for this purpose, but after working through the `Future` code it made sense to just lean into the connection with `std::future` and `std::shared_future`. I tried to call out in the comments that it should generally only be used for these caching/deduping use-cases.

File: vendor/bat-native-ledger/src/bat/ledger/internal/core/bat_ledger_initializer.h (hunk 3 above)
Comment (zenparsing) [reply vendor/bat-native-ledger/src/bat/ledger/internal/core/bat_ledger_initializer.h:1]: Fixed.

File: vendor/bat-native-ledger/src/bat/ledger/internal/core/bat_ledger_initializer.cc (hunk 1 above)
Comment (zenparsing) [reply vendor/bat-native-ledger/src/bat/ledger/internal/core/bat_ledger_initializer.cc:34]: Probably!

**Review by emerick — APPROVED** [review]
LGTM

File: vendor/bat-native-ledger/src/bat/ledger/internal/core/shared_future.h [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+// asynchronous operation is revealed to continuations as a const reference.
+// Always prefer |Future| to |SharedFuture|. In general, |SharedFuture| should
+// only be used when the result needs to be cached or deduped.
+//
+// Example:
+//   Future<int> future = MakeReadyFuture(42);
+//   SharedFuture shared_future(std::move(future));
+//   shared_future.Then(base::BindOnce([](const int& value) {}));
+template <typename T>
+class SharedFuture {
```
Comment (bridiver) [inline vendor/bat-native-ledger/src/bat/ledger/internal/core/shared_future.h:95]: I know I commented on this exact class in another PR or in DM specifically about not adding new callback abstractions

**Pull request merged**